client/src/main/java/io/pravega/client/stream/impl/ReaderGroupState.java

```diff
@@ -238,6 +246,16 @@ boolean isCheckpointComplete(String checkpointId) {
         return checkpointState.getPositionsForCompletedCheckpoint(checkpointId);
     }
 
+    @Synchronized
+    Optional<Map<Stream, StreamCut>> getStreamCutsForCompletedCheckpoint(final String checkpointId) {
+        final Optional<Map<Segment, Long>> positionMap = Optional.ofNullable(checkpointState.getPositionsForCompletedCheckpoint(checkpointId));
+
+        return positionMap.map(map -> map.entrySet().stream()
```

Comment:
How do I know that this is actually producing a valid set of stream cuts?

Reply:
Does this have good tests?

Reply:
Yes, once issue 2572 is fixed we should be adding a validation here.

> Can you argue that we are able to always produce valid stream cuts in the absence of controller validation?

Yes. It is a valid StreamCut since we are computing a StreamCut by
```
- fetching the latest read offsets of all the assigned segments. 
- using the unassigned segments.
```

Reply:
> Does this have good tests?
I have added tests for single scope multi reader scenario and multiple scope,  multi reader,  multi segment scenarios @ ReaderGroupStateTest